### PR TITLE
Feature middlewares

### PR DESCRIPTION
**Pull request by veewee**
This PR adds middleware support to the soap-client so that it is easy to configure extensions. 
More information can be found in #21.

Task list:
- [x] Specify a middleware layer
- [x] Apply handlers for triggering the SOAP requests
- [x] Adjusted the Soap-client to work with handlers
- [x] Added a Soap handler
- [x] Added a Guzzle handler
- [x] Changed the client builder to specify your own handler.
- [x] Changed the client builder to specify your own middlewares.
- [x] Created a http-binding based on https://github.com/meng-tian/soap-http-binding
- [x] Add NTLM middleware
- [x] Add Basic Auth middleware
- [x] Add WSA middleware
- [x] Add WSSE middleware
- [x] Added an easy way to load, read and change SOAP XML before sending or after receiving it.
- [x] Add spec + integration tests
- [x] Find a good way to test the middlewares
- [x] Make sure the debug() method returns correct values when using the GuzzleHandler
- [x] Add documentation
- [x] Normalize exception handling
- [x] Cleanup the codebase
- [x] End-2-End test with WSA + WSSE

Knows limitations:
- Wsse encryption does not work on newer PHP versions (#24)
- This implementation uses Guzzle as a hardcoded PHP client (#25)

### CODE REVIEW

File: src/Phpro/SoapClient/Soap/HttpBinding/Builder/Psr7RequestBuilder.php [hunk 1]
```diff
@@ -0,0 +1,234 @@
...
+
+        try {
+            $request = $this->requestFactory
+                ->createRequest($this->httpMethod, $this->endpoint)
+                ->withBody($this->prepareMessage());
+
+            foreach ($this->prepareHeaders() as $name => $value) {
+                $request = $request->withHeader($name, $value);
+            }
+        } catch (\InvalidArgumentException $e) {
```
Comment (janvernieuwe) [inline src/Phpro/SoapClient/Soap/HttpBinding/Builder/Psr7RequestBuilder.php]: use it, don't use fqcn

File: src/Phpro/SoapClient/Soap/HttpBinding/Builder/Psr7RequestBuilder.php [hunk 2]
```diff
@@ -0,0 +1,234 @@
...
+    /**
+     * @var string
+     */
+    private $httpMethod = 'POST';
+
+    /**
+     * @var RequestFactoryInterface
+     */
+    private $requestFactory;
+    /**
```
Comment (janvernieuwe) [inline src/Phpro/SoapClient/Soap/HttpBinding/Builder/Psr7RequestBuilder.php]: Add newline for max style points

File: src/Phpro/SoapClient/Soap/HttpBinding/Builder/Psr7RequestBuilder.php [hunk 3]
```diff
@@ -0,0 +1,234 @@
...
+        $headers = [];
+        $headers['Content-Length'] = (string) $this->soapMessage->getSize();
+        $headers['SOAPAction'] = $this->soapAction;
+        $headers['Content-Type'] = 'text/xml; charset="utf-8"';
+
+        return $headers;
+    }
+
+    /**
+     * SOSPAction header is removed in SOAP 1.2 and now expressed as a value of
```
Comment (janvernieuwe) [inline src/Phpro/SoapClient/Soap/HttpBinding/Builder/Psr7RequestBuilder.php]: Typo?
Also the link doesn't work for me

File: src/Phpro/SoapClient/Xml/SoapXml.php [hunk 4]
```diff
@@ -0,0 +1,159 @@
...
+use Zend\Diactoros\Stream;
+
+/**
+ * Class SoapXml
+ *
+ * @package Phpro\SoapClient\Xml
+ */
+class SoapXml
+{
+    const XMLNS_XMLNS = 'http://www.w3.org/2000/xmlns/';
```
Comment (janvernieuwe) [inline src/Phpro/SoapClient/Xml/SoapXml.php]: This looks a bit weird, is this by design?

File: src/Phpro/SoapClient/Middleware/Middleware.php [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+    }
+
+    /**
+     * @param callable $handler
+     *
+     * @return \Closure
+     */
+    public function __invoke(callable $handler)
+    {
+        return (function (RequestInterface $request, array $options) use ($handler) {
```
Comment (veewee) [inline src/Phpro/SoapClient/Middleware/Middleware.php]: Not really. This is basically the regular implementation of guzzle middlewares: http://docs.guzzlephp.org/en/latest/handlers-and-middleware.html#middleware
I made this class to make it easy to implement your own middleware, but the complexity has to be placed somewhere.

File: src/Phpro/SoapClient/ClientBuilder.php [hunk 6]
```diff
@@ -124,13 +138,44 @@ public function addTypeConverter(TypeConverterInterface $typeConverter)
...
     public function build()
     {
         $soapClientFactory = new SoapClientFactory($this->classMaps, $this->converters);
         $soapClient = $soapClientFactory->factory($this->wsdl, $this->soapOptions);
 
+        if ($this->handler) {
+            $soapClient->setHandler($this->handler);
+        }
+
+        if (count($this->middlewares)) {
```
Comment (veewee) [inline src/Phpro/SoapClient/ClientBuilder.php]: I do agree on your argument. I've just changed it, but it was more readible as it is right now. I'll keep this one AS-IS for now.

File: src/Phpro/SoapClient/Middleware/WsseMiddleware.php [hunk 7]
```diff
@@ -0,0 +1,233 @@
...
+
+    /**
+     * @var string
+     */
+    private $privateKeyFile = '';
+
+    /**
+     * @var string
+     */
+    private $publicKeyFile = '';
```
Comment (veewee) [inline src/Phpro/SoapClient/Middleware/WsseMiddleware.php]: Fixed

File: src/Phpro/SoapClient/Middleware/WsseMiddleware.php [hunk 8]
```diff
@@ -0,0 +1,233 @@
...
+    /**
+     * @param string      $username
+     * @param string|null $password
+     * @param bool        $digest
+     *
+     * @return $this
+     */
+    public function withUserToken(string $username, string $password = null, $digest = false)
+    {
+        $this->userToken = true;
```
Comment (veewee) [inline src/Phpro/SoapClient/Middleware/WsseMiddleware.php]: Fixed

File: src/Phpro/SoapClient/Middleware/WsseMiddleware.php [hunk 9]
```diff
@@ -0,0 +1,233 @@
...
+     */
+    public function afterResponse(ResponseInterface $response)
+    {
+        if (!$this->encrypt) {
+            return $response;
+        }
+
+        $xml = SoapXml::fromStream($response->getBody());
+        $wsse = new WSSESoap($xml->getXmlDocument());
+        $wsse->decryptSoapDoc($xml->getXmlDocument(), [
```
Comment (veewee) [inline src/Phpro/SoapClient/Middleware/WsseMiddleware.php]: Fixed

File: src/Phpro/SoapClient/Soap/HttpBinding/Builder/Psr7RequestBuilder.php (hunk 2 above)
Comment (veewee) [inline src/Phpro/SoapClient/Soap/HttpBinding/Builder/Psr7RequestBuilder.php]: Fixed

File: src/Phpro/SoapClient/Soap/HttpBinding/Builder/Psr7RequestBuilder.php (hunk 1 above)
Comment (veewee) [inline src/Phpro/SoapClient/Soap/HttpBinding/Builder/Psr7RequestBuilder.php]: Fixed

File: src/Phpro/SoapClient/Soap/HttpBinding/Builder/Psr7RequestBuilder.php (hunk 3 above)
Comment (veewee) [inline src/Phpro/SoapClient/Soap/HttpBinding/Builder/Psr7RequestBuilder.php]: Fixed

File: src/Phpro/SoapClient/Xml/SoapXml.php (hunk 4 above)
Comment (veewee) [inline src/Phpro/SoapClient/Xml/SoapXml.php]: Yes it is by design: The XLMNS of alias XMLNS is the url.
This way, you can e.g. also add: `XMLNS_WSA = 'url';` in a future version.

File: spec/Phpro/SoapClient/Soap/HttpBinding/LastRequestInfoSpec.php [hunk 10]
```diff
@@ -0,0 +1,93 @@
...
+        $result->getLastRequestHeaders()->shouldBe('X-Request-Header: value');
+        $result->getLastRequest()->shouldBe('REQUESTBODY');
+        $result->getLastResponseHeaders()->shouldBe('X-Response-Header: value');
+        $result->getLastResponse()->shouldBe('RESPONSEBODY');
+    }
+
+    function it_can_load_from_psr7_request_and_response_without_body()
+    {
+        $request = new Request('/', 'GET', 'php://temp', ['x-request-header' => 'value']);
+        $respone = new Response('php://memory', 204, ['x-response-header' => 'value']);
```
Comment (veewee) [inline spec/Phpro/SoapClient/Soap/HttpBinding/LastRequestInfoSpec.php]: Fixed

File: spec/Phpro/SoapClient/Soap/HttpBinding/LastRequestInfoSpec.php [hunk 11]
```diff
@@ -0,0 +1,93 @@
...
+        $result->getLastRequest()->shouldBe('');
+        $result->getLastResponseHeaders()->shouldBe('');
+        $result->getLastResponse()->shouldBe('');
+    }
+
+    function it_can_load_from_psr7_request_and_response()
+    {
+        $request = new Request('/', 'POST', 'php://temp', ['x-request-header' => 'value']);
+        $request->getBody()->write('REQUESTBODY');
+        $respone = new Response('php://memory', 200, ['x-response-header' => 'value']);
```
Comment (veewee) [inline spec/Phpro/SoapClient/Soap/HttpBinding/LastRequestInfoSpec.php]: Fixed

**Comment by veewee** [comment]
Thanks for the review @janvernieuwe. It feels like it's quite ready to ship! :)
About the strict type hints: currently there is still an older part in the code which was written for php 5.6. I'll add an issue to use strict types everywhere. That is out of the scope of this ticket.

**Pull request merged**